service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -855,33 +880,36 @@ protected void onCommandConnectionEstablished(final HonoConnection commandConnec
+        connections.add(Optional.ofNullable(registrationClientFactory)
                 .map(client -> client.isConnected())
                 .orElse(Future
                         .failedFuture(new ServerErrorException(
-                                HttpURLConnection.HTTP_UNAVAILABLE, "Device Registration client factory is not set")));
-        final Future<Void> credentialsCheck = Optional.ofNullable(credentialsClientFactory)
+                                HttpURLConnection.HTTP_UNAVAILABLE, "Device Registration client factory is not set"))));
+        connections.add(Optional.ofNullable(credentialsClientFactory)
                 .map(client -> client.isConnected())
                 .orElse(Future.failedFuture(new ServerErrorException(
-                        HttpURLConnection.HTTP_UNAVAILABLE, "Credentials client factory is not set")));
-        final Future<Void> messagingCheck = Optional.ofNullable(downstreamSenderFactory)
+                        HttpURLConnection.HTTP_UNAVAILABLE, "Credentials client factory is not set"))));
+        connections.add(Optional.ofNullable(downstreamSenderFactory)
                 .map(client -> client.isConnected())
                 .orElse(Future.failedFuture(new ServerErrorException(
-                        HttpURLConnection.HTTP_UNAVAILABLE, "Messaging client is not set")));
-        final Future<Void> commandCheck = Optional.ofNullable(commandConsumerFactory)
+                        HttpURLConnection.HTTP_UNAVAILABLE, "Messaging client is not set"))));
+        connections.add(Optional.ofNullable(commandConsumerFactory)
                 .map(client -> client.isConnected())
                 .orElse(Future.failedFuture(new ServerErrorException(
-                        HttpURLConnection.HTTP_UNAVAILABLE, "Command & Control client factory is not set")));
-        final Future<Void> deviceConnectionCheck = Optional.ofNullable(deviceConnectionClientFactory)
-                .map(client -> client.isConnected())
-                .orElse(Future.failedFuture(new ServerErrorException(
-                        HttpURLConnection.HTTP_UNAVAILABLE, "Device Connection client factory is not set")));
-        return CompositeFuture.all(tenantCheck, registrationCheck, credentialsCheck, messagingCheck, commandCheck,
-                deviceConnectionCheck).mapEmpty();
+                        HttpURLConnection.HTTP_UNAVAILABLE, "Command & Control client factory is not set"))));
+        if (deviceConnectionClientFactory instanceof DeviceConnectionClientFactory) {
+            connections.add(Optional.ofNullable(deviceConnectionClientFactory)
+                    .map(client -> ((DeviceConnectionClientFactory) client).isConnected())
+                    .orElse(Future.failedFuture(new ServerErrorException(
+                            HttpURLConnection.HTTP_UNAVAILABLE, "Device Connection client factory is not set"))));
```

Comment:
Block can be simplified since `deviceConnectionClientFactory` can't be null here.